### ISSUE TITLE
Performance Issues

### ISSUE THREAD

**Issue by BrainInBlack** [emipa606/CentralizedClimateControl#1]
Recently i noticed the following performance issues with this mod:

* ~50ms lockup while building pipes (pawn action)
* Up to 5 second lockup after saving the game
* General hefty performance hit when a network was build

The game runs just fine before placing anything from this mod, but as soon as a single piece gets placed in the world the issues listed above begin. If it was an issue that build up over the course of the network getting more complex, then it would be much of a problem, but this happens when a single tile is placed in the world hinting at a fundamental issue with the mod.

Sadly i am not familiar with Rimworld modding, otherwise i could give you a better report than this. But if you need any additional information, let me know.

Greeting BrainInBlack

**Comment by emipa606** [comment]
I have not seen this performance issues in my games and there are users reporting that their games are running fine. It seems to me that this is connected to another mod since it only affects some users.
If you can recreate this in connection with another mod I can take a look at what could be done to fix it.

**Comment by BrainInBlack** [comment]
Gonna run some test with just this mod later today, hope i can give you some more information that way.

And maybe it is just a strange interaction with another mod that is causing the issues, and CCC just emphasises the issue.

**Comment by BrainInBlack** [comment]
Turns out that Rimworld really doesn't like high refresh rates. Turned down my display to 60 and no more issues, at least nothing major.

TL;DR: It was a me issue

**Comment by BrainInBlack** [comment]
Sadly, i have to re-open this one. Though the previous statement was somewhat true, the performance issues i observed are back with a larger network. It looks like the issue is caused by the climate controller and intake fans. As soon as i remove those from the map (debug mode) the performance goes back to a state i expect at this point in the game.

To me it looks like the mod is doing a ton of stuff right after saving the game. And in general there are quite a few things going on that cause hitches and even complete lockups when alt+tabbing in and out of the game.

Just by glancing over the code, i noticed quite a few things that can be improved. Gonna setup an Env myself and make a PR as soon i took care of the obvious stuff.

**Comment by emipa606** [comment]
Very thankful for all help regarding this!

**Comment by BrainInBlack** [comment]
There already is a small improvement after removing single use local variables and useless method calls. The calls where just changing a single variable in the object and that variable had public access anyways, so no need to raise a call for that.

The local variables on the other hand brought most of the improvement in performance. A lot of them where just there for readability, but one can achieve the same thing by just formatting the code the right way.

Next up I'll take apart the network discovery algorithm, i think that there is the most performance we can gain. Also some other parts look a bit fishy to me, but I would have to dive a little deeper to see if my guess was correct.

Anyways, PR comes when i am done with the network discovery stuff. And maybe i find some other improvements along the way.

**Comment by emipa606** [comment]
Great job, looking forward to the PR!
If you can add me on steam ill put you as a contributor to the mod as well.
